lib/puppet/http/client.rb

```diff
@@ -0,0 +1,210 @@
+class Puppet::HTTP::Client
+  def initialize(pool: Puppet::Network::HTTP::Pool.new, ssl_context: nil, redirect_limit: 10, retry_limit: 100)
+    @pool = pool
+    @default_headers = {
+      'X-Puppet-Version' => Puppet.version,
+      'User-Agent' => Puppet[:http_user_agent],
+    }.freeze
+    @default_ssl_context = ssl_context
+    @resolvers = build_resolvers
+    @redirector = Puppet::HTTP::Redirector.new(redirect_limit)
+    @retry_after_handler = Puppet::HTTP::RetryAfterHandler.new(retry_limit, Puppet[:runinterval])
+  end
+
+  def create_session
+    Puppet::HTTP::Session.new(self, @resolvers)
+  end
+
+  def connect(uri, ssl_context: nil, &block)
+    ctx = ssl_context ? ssl_context : default_ssl_context
+    site = Puppet::Network::HTTP::Site.from_uri(uri)
+    verifier = Puppet::SSL::Verifier.new(site.host, ctx)
+
+    @pool.with_connection(site, verifier) do |http|
+      if block_given?
+        handle_post_connect(uri, http, &block)
+      end
+    end
+  rescue Puppet::HTTP::HTTPError
+    raise
+  rescue => e
+    raise Puppet::HTTP::ConnectionError.new(_("Failed to connect to %{uri}: %{message}") % {uri: uri, message: e.message}, e)
+  end
+
+  def get(url, headers: {}, params: {}, ssl_context: nil, user: nil, password: nil, &block)
+    query = encode_params(params)
+    unless query.empty?
+      url = url.dup
+      url.query = query
+    end
```

Comment:
What if url already contains a query. Should we merge queries, or warn that the existing query is being overwritten?